pkg/chunk/aws_storage_client.go

```diff
@@ -331,28 +340,257 @@ func (a dynamoDBRequestAdapter) HasNextPage() bool {
 	var resp *s3.GetObjectOutput
 	err := instrument.TimeRequestHistogram(ctx, "S3.GetObject", s3RequestDuration, func(ctx context.Context) error {
 		var err error
 		resp, err = a.S3.GetObjectWithContext(ctx, &s3.GetObjectInput{
 			Bucket: aws.String(a.bucketName),
-			Key:    aws.String(key),
+			Key:    aws.String(chunk.externalKey()),
 		})
 		return err
 	})
 	if err != nil {
-		return nil, err
+		return Chunk{}, err
 	}
 	defer resp.Body.Close()
 	buf, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return nil, err
+		return Chunk{}, err
+	}
+	if err := chunk.decode(buf); err != nil {
+		return Chunk{}, err
 	}
-	return buf, nil
+	return chunk, nil
 }
 
-func (a awsStorageClient) PutChunk(ctx context.Context, key string, buf []byte) error {
+func (a awsStorageClient) getDynamoDBChunks(ctx context.Context, chunks []Chunk) ([]Chunk, error) {
+	outstanding := dynamoDBReadRequest{}
+	chunksByKey := map[string]Chunk{}
+	for _, chunk := range chunks {
+		key := chunk.externalKey()
+		chunksByKey[key] = chunk
+		tableName := a.chunkTableFor(chunk.From)
+		outstanding.Add(tableName, key, nil)
+	}
+
+	result := []Chunk{}
+	unprocessed := dynamoDBReadRequest{}
+	backoff, numRetries := minBackoff, 0
+	for outstanding.Len()+unprocessed.Len() > 0 {
+		requests := dynamoDBReadRequest{}
+		requests.TakeReqs(unprocessed, dynamoDBMaxReadBatchSize)
+		requests.TakeReqs(outstanding, dynamoDBMaxReadBatchSize)
+
+		var response *dynamodb.BatchGetItemOutput
+		err := instrument.TimeRequestHistogram(ctx, "DynamoDB.BatchGetItemPages", dynamoRequestDuration, func(ctx context.Context) error {
+			var err error
+			response, err = a.DynamoDB.BatchGetItemWithContext(ctx, &dynamodb.BatchGetItemInput{
+				RequestItems:           requests,
+				ReturnConsumedCapacity: aws.String(dynamodb.ReturnConsumedCapacityTotal),
+			})
+			return err
+		})
+
+		for _, cc := range response.ConsumedCapacity {
+			dynamoConsumedCapacity.WithLabelValues("DynamoDB.BatchGetItemPages").
+				Add(float64(*cc.CapacityUnits))
+		}
+
+		if err != nil {
+			for tableName := range outstanding {
+				recordDynamoError(tableName, err)
+			}
+			if awsErr, ok := err.(awserr.Error); !ok || awsErr.Code() != provisionedThroughputExceededException {
+				return nil, err
+			}
+		}
+
+		processedChunks, err := processChunkResponse(response, chunksByKey)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, processedChunks...)
+
+		// If there are unprocessed items, backoff and retry those items.
+		if unprocessedKeys := response.UnprocessedKeys; unprocessedKeys != nil && dynamoDBReadRequest(unprocessedKeys).Len() > 0 {
+			unprocessed.TakeReqs(unprocessedKeys, -1)
+			time.Sleep(backoff)
+			backoff = nextBackoff(backoff)
+			continue
+		}
+
+		// If we get provisionedThroughputExceededException, then no items were processed,
+		// so back off and retry all.
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == provisionedThroughputExceededException {
```

Comment:
I don't see how this code is reachable. `err` is last set by `processChunkResponse`, which doesn't do AWS stuff.

I *think* it makes more sense to put it before `processChunkResponse`, since according to the comment, no items were processed anyway and we're just going to retry them all.

At that point, it might as well go inside the `if err != nil` block that's starting on what's currently line 463.